### PR TITLE
Rework cmd type : Now use an AST and a context record

### PR DESCRIPTION
**Pull request by tmarti2**
Ok this PR changes many things in the code but it should not affect the user.
At start we just wanted to work on your `TODO : Refactor to use a record `, but we thought that the code was not really trivial to understand/read.
We now have a type `context` which contains all informations needed to run a command except for the command itself (prog + args). The type `cmd` is now used to create a command using an AST.
Creating a process will create a node `Process (prog, args)`, and the same apply for all operators (`Pipe (cmd, cmd)`, `And (cmd, cmd)`, `WOut (str, cmd)` (for write_stdout_to) etc...), with a new node for each one of them. Then using run/collect/... commands, we evaluate the command AST with a given context.
As I said, I kept old functions which now only create nodes.

What do you think ?

### CODE REVIEW

File: feather.ml [hunk 1]
```diff
@@ -30,15 +30,29 @@ type env = (string * string) list
...
-  background:bool ->
-  cwd:string option ->
-  env:env option ->
-  unit
+  | Process of string * string list
+  | Pipe of cmd * cmd
+  | And of cmd * cmd
+  | Or of cmd * cmd
+  | Seq of cmd * cmd
+  | WOut of string * cmd
```
Comment (charlesetc) [inline feather.ml:48]: Do you mind using longer names here? I find abbreviations confusing sometimes.

**Review by charlesetc — COMMENTED** [review]
I think I like this! I have to read a bit more carefully still but overall I do think it helps the readability/structure of the code. I guess one thing that comes to mind is that maybe constructing an AST impacts performance negatively somehow? But my guess is the overhead from starting a process is more significant than anything we do in the internal representation. Any thoughts about that?

Thanks for the changes!

File: feather.ml (hunk 1 above)
Comment (tmarti2) [reply feather.ml:48]: No problem I can change this.

**Comment by charlesetc** [comment]
That all sounds good to me. Thanks for the due diligence regarding performance and overall a great PR!

**Pull request merged**